## Data Cleaning

In [1]:
import pandas as pd

In [47]:
df = pd.read_csv('unclean_smartwatch_health_data.csv')

In [48]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   User ID                 9799 non-null   float64
 1   Heart Rate (BPM)        9600 non-null   float64
 2   Blood Oxygen Level (%)  9700 non-null   float64
 3   Step Count              9900 non-null   float64
 4   Sleep Duration (hours)  9850 non-null   str    
 5   Activity Level          9800 non-null   str    
 6   Stress Level            9800 non-null   str    
dtypes: float64(4), str(3)
memory usage: 811.5 KB


In [4]:
df.isna().sum()

User ID                   201
Heart Rate (BPM)          400
Blood Oxygen Level (%)    300
Step Count                100
Sleep Duration (hours)    150
Activity Level            200
Stress Level              200
dtype: int64

In [14]:
df.head()

,User ID,Heart Rate (BPM),Blood Oxygen Level (%),Step Count,Sleep Duration (hours),Activity Level,Stress Level
0,4174.0,58.939776,98.809650,5450.390625,7.167235622316564,Highly Active,1
1,NaN,NaN,98.532195,727.601624,6.538239375570314,Highly_Active,5
2,1860.0,247.803052,97.052954,2826.521973,ERROR,Highly Active,5
3,2294.0,40.000000,96.894213,13797.337891,7.367789630207228,Actve,3
4,2130.0,61.950165,98.583797,15679.067383,NaN,Highly_Active,6


In [12]:
df.describe()

,User ID,Heart Rate (BPM),Blood Oxygen Level (%),Step Count
count,9799.000000,9600.000000,9700.000000,9900.000000
mean,3007.480253,76.035462,97.841581,6985.685547
std,1150.581542,19.412483,1.732863,6885.809570
min,1001.000000,40.000000,90.791208,0.910138
25%,1997.500000,64.890152,96.662683,2021.039673
50%,2998.000000,75.220601,98.010642,4962.534668
75%,4004.000000,85.198249,99.376179,9724.903320
max,4999.000000,296.593970,100.000000,62486.691406


In [18]:
df['Step Count'] = df['Step Count'].astype('float32')

In [21]:
df.duplicated().sum()

np.int64(0)

In [22]:
df.isnull().sum()

User ID                   201
Heart Rate (BPM)          400
Blood Oxygen Level (%)    300
Step Count                100
Sleep Duration (hours)    150
Activity Level            200
Stress Level              200
dtype: int64

In [23]:
df.isna().sum()

User ID                   201
Heart Rate (BPM)          400
Blood Oxygen Level (%)    300
Step Count                100
Sleep Duration (hours)    150
Activity Level            200
Stress Level              200
dtype: int64

## Heart Rate max = 296.59 → probably an outlier/error.
## Step Count max = 62,486 → potentially an outlier, depending on what the dataset represents.
## Step Count min = 0.91 → suspicious if this is daily step count.
## Blood Oxygen looks reasonable: 90.79–100.

## Changing Sleep Duration, user_id in numeric.
## Changing other numerical columns to decimal upto 2 values.

In [49]:
df['Sleep Duration (hours)'] = pd.to_numeric(df['Sleep Duration (hours)'], errors='coerce')

In [ ]:
df['Heart Rate (BPM)'].plot(kind='box')

In [32]:
df.dropna(subset=['Heart Rate (BPM)'], inplace=True)

In [33]:
df.isna().sum()

User ID                   195
Heart Rate (BPM)            0
Blood Oxygen Level (%)    289
Step Count                 94
Sleep Duration (hours)    382
Activity Level            191
Stress Level              193
dtype: int64

In [50]:
df = df[['Heart Rate (BPM)','Blood Oxygen Level (%)','Step Count','Sleep Duration (hours)', 'Activity Level', 'Stress Level']]

In [89]:
numerical_col = ['Heart Rate (BPM)','Blood Oxygen Level (%)','Step Count','Sleep Duration (hours)']

In [92]:
for col in numerical_col:
    df[col] = df[col].astype('float32')

In [54]:
for col in numerical_col:
    df[col] = df[col].fillna(df[col].median())

In [62]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Heart Rate (BPM)        10000 non-null  float32
 1   Blood Oxygen Level (%)  10000 non-null  float32
 2   Step Count              10000 non-null  float32
 3   Sleep Duration (hours)  10000 non-null  float32
 4   Activity Level          9800 non-null   str    
 5   Stress Level            9951 non-null   float32
dtypes: float32(5), str(1)
memory usage: 363.9 KB


In [94]:
df['Activity Level'] = df['Activity Level'].astype('category')

In [67]:
df['Stress Level'] = df['Stress Level'].fillna(5, inplace=True)

C:\Users\ManishAgrahari\AppData\Local\Temp\ipykernel_2732\192444406.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['Stress Level'] = df['Stress Level'].fillna(5, inplace=True)


In [58]:
df['Stress Level'] = pd.to_numeric(df['Stress Level'], errors='coerce')

In [69]:
df['Stress Level'] = df['Stress Level'].astype('int32')

In [95]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   Heart Rate (BPM)        10000 non-null  float32 
 1   Blood Oxygen Level (%)  10000 non-null  float32 
 2   Step Count              10000 non-null  float32 
 3   Sleep Duration (hours)  10000 non-null  float32 
 4   Activity Level          10000 non-null  category
 5   Stress Level            10000 non-null  int32   
dtypes: category(1), float32(4), int32(1)
memory usage: 205.3 KB


In [70]:
df.head()

,Heart Rate (BPM),Blood Oxygen Level (%),Step Count,Sleep Duration (hours),Activity Level,Stress Level
0,58.939777,98.809647,5450.390625,7.167236,Highly Active,1
1,75.220596,98.532196,727.601624,6.538239,Highly_Active,5
2,247.803055,97.052956,2826.521973,6.503308,Highly Active,5
3,40.000000,96.894211,13797.337891,7.367790,Actve,3
4,61.950165,98.583794,15679.067383,6.503308,Highly_Active,6


In [82]:
df['Activity Level'].unique()

<ArrowStringArray>
['Highly Active', 'Active', 'Sedentary', nan]
Length: 4, dtype: str

In [81]:
df['Activity Level'] = df['Activity Level'].replace('Actve', 'Active')

In [79]:
df['Activity Level'] = df['Activity Level'].replace('Seddentary', 'Sedentary')

In [77]:
df['Activity Level'] = df['Activity Level'].replace(
    'Highly_Active',
    'Highly Active'
)

In [74]:
df['Activity Level'] = df['Activity Level'].str.strip()

In [83]:
df['Activity Level'] = df['Activity Level'].fillna(
    df['Activity Level'].mode()[0]
)

In [96]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   Heart Rate (BPM)        10000 non-null  float32 
 1   Blood Oxygen Level (%)  10000 non-null  float32 
 2   Step Count              10000 non-null  float32 
 3   Sleep Duration (hours)  10000 non-null  float32 
 4   Activity Level          10000 non-null  category
 5   Stress Level            10000 non-null  int32   
dtypes: category(1), float32(4), int32(1)
memory usage: 205.3 KB


In [97]:
df.head(2)

,Heart Rate (BPM),Blood Oxygen Level (%),Step Count,Sleep Duration (hours),Activity Level,Stress Level
0,58.939999,98.809998,5450.390137,7.17,Highly Active,1
1,75.220001,98.529999,727.599976,6.54,Highly Active,5


In [100]:
numerical_col = ['Heart Rate (BPM)','Blood Oxygen Level (%)','Step Count','Sleep Duration (hours)']

In [101]:
for col in numerical_col:
    df[col] = df[col].round(2)

In [105]:
df['Step Count'] = df['Step Count'].round(0).astype('int32')

In [107]:
df['Heart Rate (BPM)'] = df['Heart Rate (BPM)'].round(0).astype('int32')

In [113]:
df['Blood Oxygen Level (%)'] = (
    df['Blood Oxygen Level (%)']
    .astype('float64')
    .round(2)
)

In [114]:
df.head()

,Heart Rate (BPM),Blood Oxygen Level (%),Step Count,Sleep Duration (hours),Activity Level,Stress Level
0,59,98.81,5450,7.17,Highly Active,1
1,75,98.53,728,6.54,Highly Active,5
2,248,97.05,2827,6.50,Highly Active,5
3,40,96.89,13797,7.37,Active,3
4,62,98.58,15679,6.50,Highly Active,6
